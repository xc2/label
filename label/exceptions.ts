import { build } from "./build-label";
import type { Label } from "./label";

/**
 * Exception for invalid label
 * @public
 */
export class InvalidLabel extends Error {
  readonly label: string;
  constructor(label: string | Label, message?: string) {
    label = typeof label === "string" ? label : build(label);
    super(message || `invalid label: ` + label);
    this.label = label;
  }
}

/**
 * Exception for resolving from sub-packages
 * @public
 */
export class CannotResolveFromSubPackages extends Error {
  constructor(public label: Label) {
    super("Cannot resolve from a label that includes subpackages: " + build(label));
  }
}

/**
 * Exception for no matched packages found
 * @public
 */
export class NoMatchedPackages extends Error {
  constructor(public label: Label) {
    super("Cannot find any matched packages: " + build(label));
  }
}

/**
 * Exception for no matched target found
 * @public
 */
export class NoMatchedTarget extends Error {
  constructor(public label: Label) {
    super("Cannot find any matched target: " + build(label));
  }
}
