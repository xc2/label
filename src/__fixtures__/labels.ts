export const labels = [
  "//foo/bar:wiz",
  "//foo/bar",
  "//foo/bar:all",
  "//foo/...",
  "//foo/...:all",
  "//foo/...:*",
  "//foo/...:all-targets",
  "//...",
  "//:all",
  ":foo",
  "bar:wiz",
  "bar/wiz",
  "bar:all",
  ":all",
  "...:all",
  "...",
  "bar/...:all",
];

export const invalids = ["//foo/.:a", "//foo:a:b"];
