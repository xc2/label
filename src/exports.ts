export * from "./build.js";
export * from "./exceptions.js";
export * from "./label.js";
export * from "./parse.js";
export * from "./query.js";
export * from "./resolve.js";
