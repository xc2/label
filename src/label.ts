export const SpecialTargetTypes = {
  ALL_TARGETS: Symbol.for("All Targets (Rules and Files)"),
  ALL_RULES: Symbol.for("All Rules Targets"),
} as const;

export const SpecialScopes = {
  UNKNOWN: Symbol.for("Relative so unknown"),
};
export interface Label {
  scope: string | symbol;
  package: string;
  includeSubPackages: boolean;
  target: symbol | string;
}

export function isRelative(label: Label) {
  return label.scope === SpecialScopes.UNKNOWN;
}
export function isWorkspaceScope(label: Label) {
  return label.scope === "";
}

export function clone(label: Label) {
  return { ...label };
}

export interface ExactLabel extends Label {
  scope: string;
  target: string;
  includeSubPackages: false;
}
