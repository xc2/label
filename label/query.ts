import { build } from "./build-label";
import { InvalidLabel, NoMatchedPackages, NoMatchedTarget } from "./exceptions";
import type { AbsoluteLabel, ExactLabel } from "./label";
import { parse, validateAbsolute } from "./parse";
import { resolve } from "./resolve";

/**
 * A specification for querying targets. E.g., Bazel-like, Buck-like, Non-FS Spec etc.
 * @public
 */
export interface Spec<Target> {
  /**
   * Validate a label
   * @param label - the label to validate
   * @returns a string or an `Error` for invalid label, and falsy values for valid label
   */
  validate: (label: AbsoluteLabel) => string | Error | void | undefined | null;
  /**
   * Lookup packages by label. Typically, it should return a list label for the `BUILD` files of these packages.
   * @param label - the label to lookup
   * @returns a list of matched packages presented by `ExactLabel`
   */
  lookup: (label: AbsoluteLabel) => ExactLabel[] | Promise<ExactLabel[]>;
  /**
   * Load targets from a package
   * @param pkg - the label that is returned by `lookup`
   * @returns a `<TargetName, Target>` map for the package
   */
  load: (pkg: ExactLabel) => Promise<Record<string, Target>>;
  /**
   * Extract targets from a package
   * @param targets - the map returned by `load`
   * @param target - the target name in a label
   * @returns the matched target names or `null` if not found
   */
  extract: (targets: Record<string, Target>, target: string) => string | string[] | null;
}

/**
 * Configuration for querying targets
 * @public
 */
export interface TargetQueryConfig {
  /**
   * Allow unmatched labels. If `true`, the query will not throw an error when no matched packages or targets are found.
   */
  allowUnmatched?: boolean;
}

/**
 * A query helper for targets
 * @public
 * @example
 * ```ts
 * import { TargetQuery } from "@109cafe/label";
 * import { DummySpec } from "@109cafe/dummy";
 * const query = new TargetQuery(new DummySpec(__dirname, { ignore: ["e2e/**", "dist/**"] }));
 * const targets = await query.query(["//foo:all", "-//foo:bar"]); // query all targets in foo except bar
 * ```
 */
export class TargetQuery<Target> {
  /**
   * @internal
   */
  allowUnmatched: boolean;

  /**
   * The spec for querying targets
   * @param spec - the spec of build system
   * @param config - configuration for `TargetQuery`
   */
  constructor(
    /**
     * @internal
     */
    public readonly spec: Spec<Target>,
    config: TargetQueryConfig = {}
  ) {
    this.allowUnmatched = config.allowUnmatched || false;
  }

  /**
   * Query targets by labels
   * @param _labels - the labels to query, with `-` prefix for excluding labels
   * @param _base - the base label to resolve the labels, must be an absolute label
   * @throws {@link InvalidLabel} if the label is invalid
   * @throws {@link NoMatchedPackages} if no matched packages are found
   * @throws {@link NoMatchedTarget} if no matched targets are found
   * @returns a `<TargetName, Target>` map of matched targets
   * @public
   */
  async query(_labels: string[], _base = "//"): Promise<Record<string, Target>> {
    const base = validateAbsolute(_base);
    const filters = _labels.map((v) => {
      if (v.startsWith("-")) {
        return { label: resolve(base, parse(v.slice(1))), negative: true };
      }
      return { label: resolve(base, parse(v)), negative: false };
    });
    for (const { label } of filters) {
      const validated = this.spec.validate(label);
      if (validated instanceof Error) {
        throw validated;
      } else if (typeof validated === "string") {
        throw new InvalidLabel(label, validated);
      }
    }
    const map: Record<
      string,
      {
        buildfile: ExactLabel;
        includes: string[];
        excludes: string[];
      }
    > = {};
    for (const { label, negative } of filters) {
      const packages = await this.spec.lookup(label);
      if (packages.length === 0 && !this.allowUnmatched) {
        throw new NoMatchedPackages(label);
      }
      for (const pkg of packages) {
        const sl = build(pkg);

        if (!map[sl]) {
          map[sl] = { buildfile: pkg, includes: [], excludes: [] };
        }
        const list = negative ? map[sl].excludes : map[sl].includes;
        list.push(label.target);
      }
    }
    const results: Record<string, Target> = {};
    for (const { buildfile, includes, excludes } of Object.values(map)) {
      const targets = await this.spec.load(buildfile);
      const extract = (target: string) => {
        const r = this.spec.extract(targets, target);
        if (!r?.length && !this.allowUnmatched) {
          throw new NoMatchedTarget({ ...buildfile, target });
        }
        return (Array.isArray(r) ? r : [r]).filter(Boolean) as string[];
      };
      const filtered = filterTarget(extract, includes, excludes);
      for (const target of filtered) {
        results[build({ ...buildfile, target })] = targets[target];
      }
    }
    return results;
  }
}
function filterTarget(
  getTargets: (target: string) => string[],
  includes: string[],
  excludes: string[]
) {
  const result = new Set<string>();
  for (const inc of includes) {
    const targets = getTargets(inc);

    for (const target of targets) {
      result.add(target);
    }
  }
  for (const f of excludes) {
    const targets = getTargets(f);
    for (const target of targets) {
      result.delete(target);
    }
  }
  return Array.from(result);
}
