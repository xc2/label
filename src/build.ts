import { type Label, SpecialTargetTypes } from "./label.js";

export function build(parsed: Label): string {
  let label = "";
  if (typeof parsed.scope === "string") {
    label += `${parsed.scope}//`;
  }

  label += [parsed.package || "", parsed.includeSubPackages ? "..." : ""].filter(Boolean).join("/");
  if (parsed.target === SpecialTargetTypes.ALL_TARGETS) {
    label += ":*";
  } else if (parsed.target === SpecialTargetTypes.ALL_RULES) {
    label += ":all";
  } else if (parsed.target && typeof parsed.target === "string") {
    label += `:${parsed.target}`;
  }
  return label;
}
