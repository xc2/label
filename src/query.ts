import { NoMatchedPackages, NoMatchedTarget } from "./exceptions.js";
import { type ExactLabel, type Label, SpecialTargetTypes } from "./label.js";
import { parse } from "./parse.js";

export interface PackageLookupResult {
  scope: string;
  package: string;
  buildfile: string;
}
export interface TargetFilter {
  /** all named targets to include */
  exactly: string[];
  /** -1: exclude all files/rules, 0: no file/rule filter, 1: include all files/rules */
  comprehensively: Record<"files" | "rules", -1 | 0 | 1>;
}

export type TargetResult =
  | { rule: unknown; file?: undefined }
  | { file: unknown; rule?: undefined };

export interface TargetQueryConfig {
  allowUnmatched?: boolean;
  lookupPackages: (label: Label) => PackageLookupResult[] | Promise<PackageLookupResult[]>;
  lookupTargets: (
    pkg: PackageLookupResult,
    filter: TargetFilter
  ) => Promise<({ target: string } & TargetResult)[]>;
}
export class TargetQuery {
  constructor(public config: TargetQueryConfig) {}
  async query(_labels: string[]): Promise<({ label: ExactLabel } & TargetResult)[]> {
    const filters = _labels.map((v) => {
      if (v.startsWith("-")) {
        return { label: parse(v.slice(1)), negative: true };
      }
      return { label: parse(v), negative: false };
    });
    const map: Record<
      string,
      { pkg: PackageLookupResult; includes: (string | symbol)[]; excludes: (string | symbol)[] }
    > = {};
    for (const { label, negative } of filters) {
      const packages = await this.config.lookupPackages(label);
      if (packages.length === 0 && !this.config.allowUnmatched) {
        throw new NoMatchedPackages(label);
      }
      for (const pkg of packages) {
        const sl = `${pkg.scope}//${pkg.package}`;

        if (!map[sl]) {
          map[sl] = { pkg, includes: [], excludes: [] };
        }
        const list = negative ? map[sl].excludes : map[sl].includes;
        list.push(label.target);
      }
    }
    const results: ({ label: ExactLabel } & TargetResult)[] = [];
    for (const { pkg, includes, excludes } of Object.values(map)) {
      const filter = getFilter({ includes, excludes });
      const targets = (await this.config.lookupTargets(pkg, filter)).map((v) => ({
        ...v,
        label: { ...pkg, target: v.target, includeSubPackages: false },
      }));
      for (const { label, rule, file } of targets) {
        if (rule || file) {
          results.push({ label, rule, file } as any);
        } else if (!this.config.allowUnmatched) {
          throw new NoMatchedTarget(label);
        }
      }
    }
    return results;
  }
}

function parseFilters(targets: (string | symbol)[]) {
  const known = new Set<string>();
  let allFiles = false,
    allRules = false;
  for (const v of targets) {
    if (v === SpecialTargetTypes.ALL_TARGETS) {
      allFiles = true;
      allRules = true;
    } else if (v === SpecialTargetTypes.ALL_RULES) {
      allRules = true;
    } else if (typeof v === "string") {
      known.add(v);
    }
  }
  return { known, allFiles, allRules };
}
function getFilter({
  includes,
  excludes,
}: { includes: (string | symbol)[]; excludes: (string | symbol)[] }): TargetFilter {
  const inc = parseFilters(includes);
  const exc = parseFilters(excludes);

  const files = exc.allFiles ? -1 : inc.allFiles ? 1 : 0;
  const rules = exc.allRules ? -1 : inc.allRules ? 1 : 0;

  for (const v of exc.known) {
    inc.known.delete(v);
  }
  return { exactly: Array.from(inc.known), comprehensively: { files, rules } };
}
