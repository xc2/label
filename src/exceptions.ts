import { build } from "./build.js";
import type { Label } from "./label.js";

export class InvalidLabel extends Error {
  readonly label: string;
  constructor(label: string | Label, message?: string) {
    label = typeof label === "string" ? label : build(label);
    super(message || `invalid label: ` + label);
    this.label = label;
  }
}

export class CannotResolveFromSubPackages extends Error {
  constructor(public label: Label) {
    super("Cannot resolve from a label that includes subpackages: " + build(label));
  }
}

export class NoMatchedPackages extends Error {
  constructor(public label: Label) {
    super("Cannot find any matched packages: " + build(label));
  }
}

export class NoMatchedTarget extends Error {
  constructor(public label: Label) {
    super("Cannot find any matched target: " + build(label));
  }
}
