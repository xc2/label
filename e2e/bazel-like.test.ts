import { describe, expect, test } from "bun:test";
import * as NodePath from "node:path";
import { InvalidLabel, NoMatchedPackages, NoMatchedTarget } from "../src/exceptions";
import { TargetQuery } from "../src/query";
import { BazelLikeFsSpec } from "./__fixtures__/bazel-like-fs-spec";

const FixturesRoot = NodePath.resolve(__dirname, "__fixtures__");

describe("bazel-like query", () => {
  const spec = new BazelLikeFsSpec((label) =>
    label.scope === ""
      ? NodePath.resolve(FixturesRoot, "workspace-basic", label.package)
      : NodePath.resolve(FixturesRoot, label.scope, label.package)
  );

  const tq = new TargetQuery(spec);
  test("should throw if no matched packages", async () => {
    await expect(tq.query(["//"])).rejects.toThrowError(NoMatchedPackages);
  });
  test("should throw if no matched targets", async () => {
    await expect(tq.query(["//foo/bar:non-exists"])).rejects.toThrowError(NoMatchedTarget);
  });

  test("should throw if package is ... and named target is given", async () => {
    await expect(tq.query(["//...:named"])).rejects.toThrowError(InvalidLabel);
  });

  test("should match specific taraget", async () => {
    await expect(tq.query(["foo/bar:wiz"])).resolves.toMatchSnapshot();
    await expect(tq.query(["foo/bar:file.txt"])).resolves.toMatchSnapshot();
  });
  test("should match default target if target is not given", async () => {
    await expect(tq.query(["foo/bar"])).resolves.toMatchSnapshot();
  });
  test('should match all rules if target is "all"', async () => {
    await expect(tq.query(["foo/bar:all"])).resolves.toMatchSnapshot();
  });

  test("should match all rules if package is ...", async () => {
    await expect(tq.query(["//foo/..."])).resolves.toMatchSnapshot();
  });

  test.each(["all-targets", "*"])(
    'should match all rules and files if target is "%s"',
    async (target) => {
      await expect(tq.query([`foo/bar:${target}`])).resolves.toMatchSnapshot();
    }
  );
  test("should exclude specific target", async () => {
    await expect(tq.query(["//foo/bar:all", "-//foo/bar:wiz"])).resolves.toMatchSnapshot();
  });

  test("should match repo target", async () => {
    await expect(tq.query(["@foo//:ababa"])).resolves.toMatchSnapshot();
  });
});
