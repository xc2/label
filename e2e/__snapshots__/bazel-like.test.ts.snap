// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`bazel-like query should match specific taraget 1`] = `
{
  "//foo/bar:wiz": {
    "rule": "this is wiz",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should match specific taraget 2`] = `
{
  "//foo/bar:file.txt": {
    "file": "file.txt",
    "type": "file",
  },
}
`;

exports[`bazel-like query should match default target if target is not given 1`] = `
{
  "//foo/bar:bar": {
    "rule": "this is default target",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should match all rules if target is "all" 1`] = `
{
  "//foo/bar:bar": {
    "rule": "this is default target",
    "type": "rule",
  },
  "//foo/bar:wiz": {
    "rule": "this is wiz",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should match all rules if package is ... 1`] = `
{
  "//foo/bar:bar": {
    "rule": "this is default target",
    "type": "rule",
  },
  "//foo/bar:wiz": {
    "rule": "this is wiz",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should match all rules and files if target is "all-targets" 1`] = `
{
  "//foo/bar:bar": {
    "rule": "this is default target",
    "type": "rule",
  },
  "//foo/bar:file.txt": {
    "file": "file.txt",
    "type": "file",
  },
  "//foo/bar:wiz": {
    "rule": "this is wiz",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should match all rules and files if target is "*" 1`] = `
{
  "//foo/bar:bar": {
    "rule": "this is default target",
    "type": "rule",
  },
  "//foo/bar:file.txt": {
    "file": "file.txt",
    "type": "file",
  },
  "//foo/bar:wiz": {
    "rule": "this is wiz",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should exclude specific target 1`] = `
{
  "//foo/bar:bar": {
    "rule": "this is default target",
    "type": "rule",
  },
}
`;

exports[`bazel-like query should match repo target 1`] = `
{
  "@foo//:ababa": {
    "rule": "ababa",
    "type": "rule",
  },
}
`;
